lib/posix-process: Make rt_sigtimedwait stub block

michpappas opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

According to rt_sigtimedwait(2) the syscall should return a signal number on success (> 0) or -1 on failure, so the zero 
value currently returned by the stubbed implementation can be misinterpreted by applications that check against failure. Update the stub to return implementation to block and never return a signal.

jschlumpp committed:
lib/posix-process: Make rt_sitgimedwait stub block

According to rt_sigtimedwait(2) the syscall should return a signal
number on success (> 0) or -1 on failure, so the zero value currently
returned by the stubbed implementation can be misinterpreted by
applications that check against failure. Update the stub to return
implementation to block and never return a signal.

Co-authored-by: Michalis Pappas <michalis@unikraft.io>
Signed-off-by: Marco Schlumpp <marco@unikraft.io>
Signed-off-by: Michalis Pappas <michalis@unikraft.io>

andreittr commented on lib/posix-process/signal/rt_sigtimedwait.c:
```diff
@@ -64,9 +69,28 @@ UK_LLSYSCALL_R_DEFINE(int, rt_sigtimedwait,
+	if (timeout) {
+		timeout_ns = ukplat_monotonic_clock() +
+			ukarch_time_sec_to_nsec(timeout->tv_sec) +
+			timeout->tv_nsec;
```
You might consider validating `timeout` to be both [positive and in canonical form](https://github.com/unikraft/unikraft/blob/staging/lib/uktimeconv/include/uk/timeutil.h#L46) before using it to compute the deadline. I can confirm testing the behavior of `sigtimedwait` and it errors out with `-EINVAL` on a bad timeout value.
Userspace can be arbitrarily silly, and it's good to be compatible.

Similar argument can apply to the implementation that does use signals, but that's not strictly in scope :shrug: .

michpappas replied:
Good point. I will also add a commit to fix the actual implementation.

andreittr reviewed (commented):
One minor comment; plus typo in commit title: "rt_sitgimedwait" -> "rt_sigtimedwait".
(I've already fixed the PR title)

michpappas committed:
lib/posix-process/signal: Check canonical timeout on rt_sigtimedwait

Add check to ensure that the timeout value passed to rt_sigtimedwait()
is canonical.

Signed-off-by: Michalis Pappas <michalis@unikraft.io>

mogasergiu commented on lib/posix-process/signal/rt_sigtimedwait.c:
```diff
@@ -64,9 +69,30 @@ UK_LLSYSCALL_R_DEFINE(int, rt_sigtimedwait,
+		timeout_ns = ukplat_monotonic_clock() +
+			ukarch_time_sec_to_nsec(timeout->tv_sec) +
+			timeout->tv_nsec;
+		uk_thread_block_until(current, (__snsec) timeout_ns);
```
nitpick but no space after cast (i know that's just how it was initially when you picked it up)? it;s just a checkpatch `CHECK` but as you know (and as you also do) we tend to respect that :sweat_smile:

michpappas replied:
Oh yes it's from the original author, thanks for catching. I wonder why checkpatch didn't catch....

mogasergiu reviewed (approved):
just a tiny nit, feel free to disregard if you want

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented on lib/posix-process/signal/rt_sigtimedwait.c:
```diff
@@ -42,6 +47,8 @@ UK_LLSYSCALL_R_DEFINE(int, rt_sigtimedwait,
 		goto out;
 
 	if (timeout)
+		if (unlikely(!uk_time_spec_canonical(timeout))
```
Wait I think you also meant to add { } to this if/else now

michpappas replied:
Oooops

andreittr commented on lib/posix-process/signal/rt_sigtimedwait.c:
```diff
@@ -64,9 +72,30 @@ UK_LLSYSCALL_R_DEFINE(int, rt_sigtimedwait,
+	int rc;
+
+	if (timeout) {
+		if (unlikely(!uk_time_spec_canonical(timeout)))
```
We should check both canonicity _and_ a positive value (the two are independent properties, as a canonical timespec can indeed represent negative time)
```c
if (unlikely(!(uk_time_spec_canonical(timeout) && uk_time_spec_positive(timeout))))
```

michpappas replied:
Right, I got mislead by the check for `tv->tv_nsec` thinking that this would suffice. Would be useful if that was documented. Also a bit strange that negative time can only be expressed by the `sec` part but not the `nsec`, that would be also useful to be aware of.

andreittr replied:
Because the two are independent properties. A `struct timespec`, as [documented](https://www.man7.org/linux/man-pages/man3/timespec.3type.html) can only have `tv_nsec` in the range `[0, 999'999'999]` regardless of type or signedness of `tv_nsec`, and regardless of sign of `tv_sec`. We call timespec structs that follow this rule "canonical". AFAICT all Linux/libc API outputs canonical timespecs and errors out if fed non-canonical values.

Orthogonal to canonicity is the actual value of the timespec (which depending on context is interpreted either as a relative "time offset from now" or absolute "offset from epoch") that can be either positive or negative. In the case of a timeout, only positive timespec values make sense, but for timestamps negative values are perfectly reasonable (e.g., `(struct timespec){ -1, 500000000 }` represents half a second before midnight Jan 1 1970).

andreittr reviewed (approved):
Looks good, thanks!

Reviewed-by: Andrei Tatar <andrei@unikraft.io>

mogasergiu reviewed (approved):
Approved-by: Sergiu Moga <sergiu@unikraft.io>
